Add three new graphic classes and some functions for graphic tree comparison

Niko-creater opened the pull request:
1. I add three new graphic classes Above, Beside and Overlay which implemented exactly same as before with skia(path, pin_position) and some simple tests in "test_operation.py"      

2. I expose graphic attributes of all Graphic class for coming graphic tree compison use    

3. the file "graphic_tree_canvas" and the new added attributes "left_node" and "right_node" is just for making printing job of graphic tree easier and graphic tree printing is basically for testing and debugging. All these thing could be deleted at the end.                     

4.the function "get_point_name" (for graphic tree printing)seems stupid and redudent and I try to make it better but I failed. If you know how to make it better please help me with it.

Commit:
add print graphic tree function

Commit:
add some helper functions to help build and print tree structure

Commit:
add some comments for graphic_tree_canvas.py

Commit:
add new graphic class Beside

Commit:
rename tree canvas

Commit:
update beside function with new Beside class

Commit:
add test for new beside function

Commit:
add test for new above function

Commit:
update above function with new Above class

Commit:
add new Above class

Commit:
add test for new overlay function

Commit:
update overlay function with new Overlay class

Commit:
add new Overlay class

Commit:
minor change

Commit:
some minor change

lucach commented:
A couple of high-level points that need to be addressed before diving into the details:
- Code that pertains to debugging/visualizing/comparing graphics should not be part of this library but of the separate project that deals with the assessment. 
- The implementation of the `__init__` and `draw` methods in `Above`/`Beside`/`Overlay` should make use of the implementations of those methods in `Compose`/`Pin`. Besides the amount of code duplication, it is quite hard to make sure that all the computations you re-implemented are correct in all edge cases. The methods should essentially mimic what was done before at the function level (i.e., the combination of pin+pin+compose+pin).

lucach commented on pytamaro/color.py:
```diff
@@ -31,3 +31,8 @@ def as_tuple(self) -> Tuple[int, int, int, float]:
+
+    # Convert a Color object to a string which is used to indicate the color information in graphic tree
+    def as_string(self) -> str:
+        return "rgb:" + str(self.color[0] * 255) + "_" + str(self.color[1] * 255) + "_" + str(self.color[2] * 255) \
```
BTW: In your separate project, look into f-strings when you need to do this kind of string interpolation.

Commit:
delete test functions

Commit:
delete unnecessary test functions

Commit:
re-implement three new Graphic class with Pin and Compose

Commit:
delete graphic_tree_canvas.py

Niko-creater commented:
Hi Luca I have modified my codes according to your feedback, please check them again. Thanks a lot!

lucach commented on pytamaro/graphic.py:
```diff
@@ -114,25 +118,35 @@ class Rectangle(Primitive):
     def __init__(self, width: float, height: float, color: Color):
         path = Path().addRect(Rect.MakeWH(width, height))
         super().__init__(path, color)
+        # For graphic tree comparison
```
Mmh. Given that these attributes are useful in general, I would simply store them at the beginning of the `__init__` method, without any specific comment ("for graphic tree comparison").

It might be possible to turn all of them into dataclasses (so that we don't need to manually perform all these assignments), but then you would probably need to fight with [Post init processing](https://docs.python.org/3/library/dataclasses.html#post-init-processing).

lucach commented on pytamaro/graphic.py:
```diff
@@ -114,25 +118,35 @@ class Rectangle(Primitive):
     def __init__(self, width: float, height: float, color: Color):
         path = Path().addOval(Rect.MakeWH(width, height))
         super().__init__(path, color)
+        # For graphic tree comparison
```
Same as above (here you can see why doing things manually is a bad idea: why is the color not stored here?)

lucach commented on pytamaro/graphic.py:
```diff
@@ -144,6 +158,10 @@ def __init__(self, radius: float, angle: float, color: Color):
             path.close()
         super().__init__(path, color)
         self.set_pin_position(radius, radius)
+        # For graphic tree comparison
```
Same as above (and in all the classes below, I won't comment further).

lucach commented on pytamaro/graphic.py:
```diff
@@ -187,13 +211,18 @@ def __init__(self, text: str, font_name: str, size: float, color: Color):
     Represents the composition of two graphics, one in the foreground and the
     other in the background, joined on their pinning positions.
     """
+
```
Why are there all these extra newlines? Have you checked whether the code conforms to the style checks (e.g., by running `pylint`)?

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
         self.graphic.path.transform(self.rot_matrix, self.path)  # updates self.path
         self.pin_position = graphic.pin_position
+        # For graphic tree comparison
+        self.degree = - deg
```
No space after unary minus (this would go away if we use a dataclass -- but then you need to update any client code).
I'd store the parameter unmodified.

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
+
+class Beside(Graphic):
+    """
+    Represent the composition of two graphics that beside with each other,
```
Represent**s** the composition of two graphics one beside the other,
vertically centered.

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
+        self.left_graphic = left_graphic
+        self.right_graphic = right_graphic
+
+        b_g = Pin(Compose(Pin(left_graphic, center_right), Pin(right_graphic, center_left)), center)
```
Can we use a more descriptive name? I assume `g` stands for graphic, `b`?

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
+        self.set_pin_position(b_g.pin_position.x(), b_g.pin_position.y())
+
+    def draw(self, canvas: Canvas):
+        b_g = Pin(Compose(Pin(self.left_graphic, center_right), Pin(self.right_graphic, center_left)), center)
```
Oh no, let's not compute this twice (and also duplicate code)!
It can be stored as an attribute.

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
+
+class Above(Graphic):
+    """
+    Represent the composition of two graphics that one above other,
```
Represents the composition of two graphics one above the other,
horizontally centered.

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
+        self.set_pin_position(a_g.pin_position.x(), a_g.pin_position.y())
+
+    def draw(self, canvas: Canvas):
+        a_b = Pin(Compose(Pin(self.top_graphic, bottom_center), Pin(self.bottom_graphic, top_center)), center)
```
Same as above. Oh, now I get it: this is short for "above graphic". What about `composed_graphic`?

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
+        self.set_pin_position(o_g.pin_position.x(), o_g.pin_position.y())
+
+    def draw(self, canvas: Canvas):
+        o_g = Pin(Compose(Pin(self.front_graphic, center), Pin(self.back_graphic, center)), center)
```
Same.

lucach commented on tests/test_graphic.py:
```diff
@@ -1,5 +1,5 @@
 from pytamaro.color_names import blue, red
-from pytamaro.primitives import empty_graphic, rectangle
+from pytamaro.primitives import empty_graphic, rectangle, ellipse
```
Why is this added?

lucach commented on pytamaro/operations.py:
```diff
@@ -107,7 +106,7 @@ def overlay(foreground_graphic: Graphic, background_graphic: Graphic) -> Graphic
     check_graphic(foreground_graphic, "foreground_graphic")
     check_graphic(background_graphic, "background_graphic")
-    return _compose_pin_center(foreground_graphic, background_graphic, center, center)
+    return Overlay(foreground_graphic, background_graphic)
```
I guess this means that the `_compose_pin_center` function is no longer needed.

lucach commented on pytamaro/graphic.py:
```diff
@@ -154,13 +172,18 @@ class Triangle(Primitive):
+        # For graphic tree comparison
+        self.side1 = side1
+        self.side2 = side2
+        self.angle = angle
```
Missing color?

lucach commented on pytamaro/graphic.py:
```diff
@@ -187,13 +211,18 @@ def __init__(self, text: str, font_name: str, size: float, color: Color):
+        # For graphic tree comparison
+        self.text = text
+        self.font_name = font_name
+        self.text_size = size
```
Missing color?

lucach commented on pytamaro/graphic.py:
```diff
@@ -242,15 +274,78 @@ class Rotate(Graphic):
+        self.front_graphic = front_graphic
+        self.back_graphic = back_graphic
+
+        o_g = Pin(Compose(Pin(front_graphic, center), Pin(back_graphic, center)), center)
```
If I'm not missing something, I think one should be able to factor out this common logic among the three classes (e.g., in an intermediate class in the hierarchy named something like `SimpleCompose` which extends `Graphic` and from which `Above` `Beside` and `Overlay` are further derived).
They could share the same `draw` method, which would make use of the shared `composed_graphic` attribute.

Commit:
correct inappropriate code

Niko-creater commented:
Hi Luca, I have corrected all the inappropriate code based on your comments and I feel sorry for bringing you such bad code. Thank you so much for providing me detailed feedback and I really enjoy learning from them. If there still exists any code need to be improved, please let me know :D

lucach commented:
Hi! No need to apologize or feel bad! :-)
We all write code that can be improved, and this is indeed an opportunity to learn!

The changes are now closer to how I imagined them. I still have a couple of doubts:
- Is there a reason behind turning only `SimpleCompose` into a dataclass (and not all the other classes, which would have spared all the assignments to store the parameters)?
- `SimpleCompose` could take care of doing the composition with the extra pin on the center (exactly like `_compose_pin_center` was abstracting the behavior of the three specialized versions), and `Above`/`Beside`/`Overlay` could just pass the relevant parameters (the graphics and the points).

Commit:
update SimpleCompose class

Niko-creater commented:
Hi Luca, I follow your second suggestion and now SimpleCompose functions like previous `_compose_pin_center`.
As for your first suggestion, I try to rewrite `Above\Overlay\Beside` three classes as dataclass, but because of the inheritance of dataclass, when we need to create a new object of one class, we have to assign all the fields of this class and its inherited classes.(for instance after I turn Beside into dataclass, the `Beside.__init__()` will required 8 positional arguments: 'pin_position', 'path'(from inherited dataclass `Graphic`), 'graphic1', 'graphic2', 'point1', 'point2'(from inherited dataclass `SimpleCompose`), 'left_graphic', and 'right_graphic'. And I try to use post-init processing and init-only variables to fix this but I fail. 
I dont know if I miss something and if you have any thoughts to solve this please share with me:D

lucach committed:
Refactor Graphic class hierarchy

- Introduce specialized classes for Beside / Above / Overlay, with a BasicCompose superclass that generalizes their behavior
- Each class should store all its parameters (which might turn useful for assessment)
- More uniform reuse of superclasses' constructors
- Use frozen dataclasses to emulate immutability
- Dataclasses on the entire hierarchy provide a nicer __repr__
- Manually implement most __init__ methods to allow deriving certain attributes automatically, without having to provide them to each constructor. Given that the classes are now frozen, attributes are set via object.__setattr__().
- Disable automatically generated __eq__ to use the custom __hash__ (given that Skia objects are unhashable)

lucach commented:
It's not your fault! Unfortunately, this whole `dataclass` approach is still rather messy for non-trivial use cases like ours. I tried to come up with a solution in https://github.com/LuCEresearchlab/pytamaro/pull/22/commits/f154990d57b0d75acc25b3a15717d1088598f7df. As you can see, it's quite a lot of changes.

The commit message explains in some detail the compromise I reached. It's probably not as clean as it could be, but it should fulfill all our requirements. Eventually, I decided to manually implement the constructors so that we could compute the required fields for the superclass(es). Among the other benefits, PyTamaro `Graphic`s now have slightly more informative string representations and ""emulate immutability"".

This should be ready to be merged, but before doing that I'd like you to try to update your graphics-comparison code, to see whether these new changes address 100% of the aspects that were missing before.

Niko-creater commented:
Hi Luca. Thanks for detailed commit messages and they are very helpful! Currently I'm still struggling with the implemetion of my new graphic tree comparison and feedback algorithm, but I can confirm that the current version pytamaro has fulfil my need. So I think the code could be merged.

lucach commented:
No news is good news... I am therefore merging this so that I can make other changes to the library that depend on this.

Niko-creater commented:
Glad to hear that:3
